torch_frame/nn/models/excelformer.py

```diff
@@ -37,30 +39,59 @@ def feature_mixup(
         y (Tensor): The target.
         num_classes (int): Number of classes.
         beta (float): The concentration parameter of the Beta distribution.
+            (default: :obj:`0.5`)
+        mixup_type (str): The mixup methods. `none` option means no mixup,
+            options `feature` and `hidden` are `FEAT-MIX` (mixup at feature
+            dimensionality) and `HIDDEN-MIX` (mixup at hidden dimensionality)
+            proposed in ExcelFormer paper. (default: :obj:`none`)
+        mi_scores (Tensor): Mutual information scores only used in the
+            mixup weight calculation for `FEAT-MIX`. (default: :obj:`None`)
 
     Returns:
         x_mixedup (Tensor): The mixedup numerical feature.
         y_mixedup (Tensor): Transformed target of size
             :obj:`[batch_size, num_classes]`
     """
     assert num_classes > 0
+    assert mixup_type in ['none', 'feature', 'hidden']
+
+    beta = torch.tensor(beta, dtype=torch.float32, device=x.device)
     beta_distribution = torch.distributions.beta.Beta(beta, beta)
     shuffle_rates = beta_distribution.sample(torch.Size((len(x), 1)))
-    feat_masks = torch.rand(x.shape, device=x.device) < shuffle_rates
     shuffled_idx = torch.randperm(len(x), device=x.device)
+    assert x.ndim == 3, 'FEAT-MIX or HIDDEN-MIX requires encoded features'
+    if mixup_type == 'feature':
+        assert mi_scores is not None
+        mi_scores = mi_scores.to(x.device)
+        # Hard masks (feature dimension)
+        feat_masks = torch.rand(torch.Size((x.shape[0], x.shape[1])),
+                                device=x.device) < shuffle_rates
+        l1_norm_mi = mi_scores / mi_scores.sum()
+        lmbd = torch.sum(
+            l1_norm_mi.unsqueeze(0) * feat_masks, dim=1, keepdim=True)
+        feat_masks = feat_masks.unsqueeze(2)
+    elif mixup_type == 'hidden':
+        # Hard masks (hidden dimension)
+        feat_masks = torch.rand(torch.Size((x.shape[0], x.shape[2])),
```

Comment:
same here. can be made simpler.
`torch.rand(x.shape, device=...)

Reply:
As mentioned above, `feature` or `hidden` options only mixup at feature or hidden dimension, `torch.rand(x.shape, device=...)` makes inconsistent mixup weights in both dimensions that is not expected in the paper, for example, using `hidden` means each feature share the same mixup weights of hidden dimension. `torch.rand(x.shape, device=...)` indeed creates a mixture of `feature` & `hidden` while this mixup manner performs not good in our trail.

Reply:
makes sense.